ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -355,15 +298,58 @@ AmbryReplica getReplicaForPartitionOnNode(String hostname, int port, String part
   }
 
   /**
-   * An instance of this object is used to register as listener for Helix related changes in each datacenter.
+   * An instance of this object is used to register as listener for Helix related changes in each datacenter. This
+   * class is also responsible for handling events received.
    */
-  private class ClusterChangeListener implements ExternalViewChangeListener, InstanceConfigChangeListener, LiveInstanceChangeListener {
+  private class ClusterChangeHandler implements InstanceConfigChangeListener, LiveInstanceChangeListener {
+    private final String dcName;
     final Set<String> allInstances = new HashSet<>();
-    private boolean liveInstanceChangeTriggered = false;
-    private boolean externalViewChangeTriggered = false;
-    private boolean configChangeTriggered = false;
     private final Object notificationLock = new Object();
-    private final CountDownLatch initialized = new CountDownLatch(3);
+    private Exception initializationException;
+
+    /**
+     * Initialize a ClusterChangeHandler in the given datacenter.
+     * @param dcName the datacenter associated with this ClusterChangeHandler.
+     */
+    ClusterChangeHandler(String dcName) {
+      this.dcName = dcName;
+    }
+
+    @Override
+    public void onInstanceConfigChange(List<InstanceConfig> configs, NotificationContext changeContext) {
```

Comment:
for the ease of reviewing, if any method has been moved around w/o any changes as such, do leave a word as git hub comment on the same. Something to keep in mind for future.